tensorflow/contrib/go/session.go

```diff
@@ -0,0 +1,147 @@
+package tensorflow
+
+import (
+	"fmt"
+	"runtime"
+
+	"github.com/golang/protobuf/proto"
+)
+
+// Session A Session instance lets a caller drive a TensorFlow graph
+// computation.
+type Session struct {
+	ops     TF_SessionOptions
+	session TF_Session
+	status  TF_Status
+	graph   *Graph
+}
+
+// NewSession initializes a new TensorFlow session.
+func NewSession() (s *Session, err error) {
+	status := TF_NewStatus()
+	ops := TF_NewSessionOptions()
+	s = &Session{
+		ops:    ops,
+		status: status,
+		session: TF_NewSession(
+			ops,
+			status,
+		),
+	}
+
+	err = s.statusToError(status)
+
+	// Release the C allocated memory when the instance is destroyed
+	runtime.SetFinalizer(s, (*Session).FreeAllocMem)
+
+	return
+}
+
+// Run Runs the operations on the target nodes, or all the operations if not
+// targets are specified. the Parameter Input in a dictionary where the key is
+// the tensor name on the graph, and the value the Tensor. The parameter
+// outputs is used to specify the tensors from the graph to be returned in the
+// same order as they occur on the slice.
+func (s *Session) Run(inputs map[string]*Tensor, outputs []string, targets []string) ([]*Tensor, error) {
+	inputNames := NewStringVector()
+	inputValues := NewTensorVector()
+	for k, v := range inputs {
+		v.setCMemAsAlreadyRelease()
+		inputValues.Add(v.tensor)
+		inputNames.Add(k)
+	}
+	outputNames := NewStringVector()
+	for _, n := range outputs {
+		outputNames.Add(n)
+	}
+
+	targetNames := NewStringVector()
+	for _, n := range targets {
+		targetNames.Add(n)
+	}
+
+	outputValues := NewTensorVector()
+	status := TF_NewStatus()
+	defer TF_DeleteStatus(status)
+
+	TF_Run_wrapper(
+		s.session,
+		inputNames,
+		inputValues,
+		outputNames,
+		outputValues,
+		targetNames,
+		status)
+
+	result := make([]*Tensor, 0, outputValues.Size())
+	for i := int64(0); i < outputValues.Size(); i++ {
+		result = append(result, &Tensor{
+			tensor: outputValues.Get(int(i)),
+		})
+	}
+
+	return result, s.statusToError(status)
+}
+
+// ExtendGraph Loads the graph definition on the session.
+func (s *Session) ExtendGraph(graph *Graph) (err error) {
+	status := TF_NewStatus()
+	defer TF_DeleteStatus(status)
+	buf, err := proto.Marshal(graph.def)
+	if err != nil {
+		return err
+	}
+
+	TF_ExtendGraph(s.session, buf, status)
+	s.graph = graph
+
+	return s.statusToError(status)
+}
+
+// ExtendAndInitializeAllVariables Adds the "init" op to the graph in order to
```

Comment:
I'll stop noting these, but would be good to search through all the function comments and de-capitalize the second word in case like this, as per comment sentences.